File: src/main/attribute.js
'use strict';

const { convert, isLLMAvailable } = require('./llm');

/**
 * Per-line artist attribution — which collaborator sings which line.
 *
 * WHY. 0.20.0 gave every artist a distinct silhouette, which fixed *identity*:
 * you can now tell the dancers apart. It did not fix *meaning*. The dancer on
 * the mic is still chosen by `line index % member count`, so on a three-artist
 * collaboration the figure singing is right one time in three by accident, and
 * the app confidently shows the wrong person for the rest of the song. A
 * feature that is wrong two thirds of the time is worse than one that makes no
 * claim at all.
 *
 * WHY AN LLM. No lyric source carries this. LRCLIB, NetEase and every other
 * free source return lines and timings; verse ownership is not in the data at
 * any price. But it is exactly the kind of thing a model knows about a released
 * song, and where it does not know, the structure of the lyrics usually gives
 * it away — a verse boundary, a name-check, a switch in voice.
 *
 * WHAT IT MUST NOT DO. This never touches the words or the timings. It returns
 * an array of indices into the artist list, one per line, and nothing else. The
 * failure mode is falling back to what 0.20.0 did.
 *
 * ONE PASS PER TRACK, CACHED. The answer is a property of the song, not of the
 * play, so it is computed once and stored beside the cues. Nothing here runs
 * for a solo artist: with one name in the list there is no question to ask.
 */

/**
 * Lines per request. Larger than correction's 40 because the model needs to see
 * whole verses to spot where one ends — and each line costs only a number back,
 * not a rewritten line.
 */
const BATCH_SIZE = 80;

/** Never ask about a lineup bigger than this; past it the answer is noise. */
const MAX_ARTISTS = 6;

const SYSTEM_PROMPT = `You identify which artist performs each line of a song.

You are given a song title, an ordered list of the artists credited on it, and
the song's lyrics as numbered lines. For every line, say which artist sings it.

Rules:
- Answer with an artist INDEX from the list you were given, zero-based.
- Return EXACTLY as many answers as there are lines, in the same order.
- Songs are usually sung in blocks, not alternating line by line. A verse
  belongs to one artist; a hook is often shared. Prefer contiguous runs.
- Use what you know about the recording first. Where you do not know it, use the
  structure: verse boundaries, a change of voice, an artist named in the line.
- If a line is sung together or you genuinely cannot tell, use -1. That is a
  real answer and is better than a guess; the app has a sensible default for it.
- Never invent an artist who is not in the list, and never return an index
  outside it.`;

const SCHEMA = {
  type: 'object',
  properties: {
    singers: {
      type: 'array',
      items: { type: 'integer' },
      description:
        'One zero-based artist index per lyric line, in order. -1 means shared or unknown.',
    },
  },
  required: ['singers'],
  additionalProperties: false,
};

/** Whether attribution can run at all (any LLM provider configured). */
function isAttributionAvailable() {
  return isLLMAvailable();
}

/**
 * Split a credit string into the individual artists.
 *
 * Deliberately the same shapes SMTC and the lyric sources actually produce —
 * `A, B & C`, `A feat. B`, `A x B`, `A vs C` — and deliberately NOT a general
 * name parser. Anything it does not recognise stays one artist, which is the
 * safe direction: a lineup that is too short asks a smaller question, where a
 * lineup that split a single name in half asks a nonsensical one.
 *
 * @param {string} credit
 * @returns {string[]}
 */
function splitArtists(credit) {
  if (typeof credit !== 'string' || !credit.trim()) return [];
  return credit
    /* The word boundary goes BEFORE the optional dot, not after it. `\bfeat\.?\b`
       cannot match the dot in "feat. 21 Savage" — there is no word character
       after it for the boundary to sit against — so it matched "feat" alone and
       left ". 21 Savage" as an artist name. */
    .split(/\s*(?:,|&|\b(?:feat|ft|featuring|with|vs|x)\b\.?|\/)\s*/i)
    .map((s) => s.trim())
    .filter(Boolean)
    /* Case-insensitive dedupe: "Drake" and "drake" are one person, and a
       repeated name would give the model two indices for one dancer. */
    .filter((name, i, all) => all.findIndex((o) => o.toLowerCase() === name.toLowerCase()) === i)
    .slice(0, MAX_ARTISTS);
}

/**
 * Validate and normalise a model's answer for one batch.
 *
 * Pure, and separate from the network call because this is where a bad response
 * would do its damage. Every index is clamped into the artist list, anything
 * unparseable becomes -1 (shared/unknown, which the renderer already handles),
 * and a wrong-length answer is rejected outright rather than silently aligning
 * the wrong lines to the wrong people.
 *
 * @param {number} lineCount
 * @param {number[]} singers
 * @param {number} artistCount
 * @returns {number[]}
 * @throws {Error} when the count does not match
 */
function normaliseSingers(lineCount, singers, artistCount) {
  if (!Array.isArray(singers) || singers.length !== lineCount) {
    throw new Error(
      `Answer count mismatch: sent ${lineCount}, received ${Array.isArray(singers) ? singers.length : 0}.`
    );
  }
  return singers.map((raw) => {
    /* Checked before Number(), because `Number(null)` is 0 and `Number(false)`
       is 0 — a missing answer would silently become "the first artist sings
       this line", which is precisely the confident-but-wrong outcome this
       whole module exists to remove. A numeric string is still accepted: that
       is a loose provider, not a missing answer. */
    if (raw === null || raw === undefined || typeof raw === 'boolean') return -1;
    const n = Number(raw);
    if (!Number.isInteger(n) || n < 0 || n >= artistCount) return -1;
    return n;
  });
}

/**
 * Smooth an attribution into blocks.
 *
 * Songs are performed in verses, not in alternation, so a lone line credited to
 * a different artist between two long runs of one is far more likely to be a
 * model slip than a real one-line handoff — and on screen it is a dancer
 * flickering onto the mic for a second and back, which reads as a bug whether
 * or not it is one. Runs shorter than `minRun` are absorbed into the run before
 * them.
 *
 * Applied after validation, on purpose: it is a presentation decision, and
 * keeping it separate means the raw answer stays testable.
 *
 * @param {number[]} singers
 * @param {number} [minRun=2]
 * @returns {number[]}
 */
function smoothRuns(singers, minRun = 2) {
  if (!Array.isArray(singers) || singers.length === 0) return [];
  const out = singers.slice();
  let start = 0;
  while (start < out.length) {
    let end = start;
    while (end + 1 < out.length && out[end + 1] === out[start]) end += 1;
    const length = end - start + 1;
    // The first run has nothing before it to be absorbed into, so it stands.
    if (length < minRun && start > 0) {
      for (let i = start; i <= end; i += 1) out[i] = out[start - 1];
    }
    start = end + 1;
  }
  return out;
}

/**
 * Work out who sings each line.
 *
 * Resolves to `null` on anything less than a usable answer — no provider, a
 * solo artist, a network failure, a malformed response. Null means "carry on
 * doing what 0.20.0 did", which is a working app, so no caller needs to handle
 * an error.
 *
 * @param {Array<{text: string}>} cues
 * @param {{title?: string, artist?: string}} track
 * @param {string[]} [artists] override the parsed lineup
 * @returns {Promise<{artists: string[], singers: number[]}|null>}
 */
async function attributeLines(cues, track, artists) {
  if (!Array.isArray(cues) || cues.length === 0) return null;
  if (!isAttributionAvailable()) return null;

  const lineup = Array.isArray(artists) && artists.length
    ? artists.slice(0, MAX_ARTISTS)
    : splitArtists((track && track.artist) || '');

  // One name is not a question. Two dancers is where the 0.20.0 behaviour
  // starts being visibly wrong, so that is where this starts earning its cost.
  if (lineup.length < 2) return null;

  const batches = Math.ceil(cues.length / BATCH_SIZE);
  const singers = [];

  for (let b = 0; b < batches; b += 1) {
    const slice = cues.slice(b * BATCH_SIZE, (b + 1) * BATCH_SIZE);
    try {
      const parsed = await convert({
        system: SYSTEM_PROMPT,
        user:
          `Song: ${(track && track.title) || 'unknown'}\n`
          + `Artists (use these indices):\n`
          + lineup.map((name, i) => `  ${i}: ${name}`).join('\n')
          + '\n\nLines:\n'
          + slice.map((c, i) => `${i}: ${(c.text || '').trim()}`).join('\n')
          + `\n\nReturn exactly ${slice.length} indices.`,
        schema: SCHEMA,
      });
      singers.push(...normaliseSingers(slice.length, parsed.singers, lineup.length));
    } catch (err) {
      /* Per batch, like correction: one bad response should not throw away the
         verses that were attributed correctly. -1 is the honest filler — it
         means "shared or unknown", which is exactly the situation. */
      console.warn(`[attribute] batch ${b + 1}/${batches} skipped:`, err.message);
      singers.push(...new Array(slice.length).fill(-1));
    }
  }

  // Every line unknown is not an attribution; it is the default wearing a hat.
  if (singers.every((s) => s < 0)) return null;

  return { artists: lineup, singers: smoothRuns(singers) };
}

module.exports = {
  attributeLines,
  splitArtists,
  normaliseSingers,
  smoothRuns,
  isAttributionAvailable,
  BATCH_SIZE,
  MAX_ARTISTS,
};

File: src-tauri/src/attribute.rs
//! Per-line artist attribution — which collaborator sings which line.
//!
//! A Rust port of `attribute.js`. No lyric source carries this, but a model
//! knows it (or can infer it from verse structure). Returns one artist index
//! per line; never touches words or timings. Falls back to None (the renderer
//! keeps rotating dancers) on anything less than a usable answer.

use std::sync::LazyLock;

use regex::Regex;
use serde_json::{json, Value};

use crate::llm;

const BATCH_SIZE: usize = 80;
const MAX_ARTISTS: usize = 6;

const SYSTEM_PROMPT: &str = "You identify which artist performs each line of a song.\n\n\
You are given a song title, an ordered list of the artists credited on it, and\n\
the song's lyrics as numbered lines. For every line, say which artist sings it.\n\n\
Rules:\n\
- Answer with an artist INDEX from the list you were given, zero-based.\n\
- Return EXACTLY as many answers as there are lines, in the same order.\n\
- Songs are usually sung in blocks, not alternating line by line. A verse\n\
  belongs to one artist; a hook is often shared. Prefer contiguous runs.\n\
- Use what you know about the recording first. Where you do not know it, use the\n\
  structure: verse boundaries, a change of voice, an artist named in the line.\n\
- If a line is sung together or you genuinely cannot tell, use -1.\n\
- Never invent an artist who is not in the list, and never return an index\n\
  outside it.";

static SPLIT_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"(?i)\s*(?:,|&|\b(?:feat|ft|featuring|with|vs|x)\b\.?|/)\s*").unwrap());

fn schema() -> Value {
    json!({
        "type": "object",
        "properties": {
            "singers": {
                "type": "array",
                "items": { "type": "integer" },
                "description": "One zero-based artist index per lyric line, in order. -1 means shared or unknown.",
            },
        },
        "required": ["singers"],
        "additionalProperties": false,
    })
}

/// Split a credit string into individual artists (the shapes SMTC/lyric sources
/// actually produce). Case-insensitive dedupe, capped at MAX_ARTISTS.
pub fn split_artists(credit: &str) -> Vec<String> {
    if credit.trim().is_empty() {
        return Vec::new();
    }
    let mut out: Vec<String> = Vec::new();
    for part in SPLIT_RE.split(credit) {
        let name = part.trim();
        if name.is_empty() {
            continue;
        }
        if out.iter().any(|o| o.eq_ignore_ascii_case(name)) {
            continue;
        }
        out.push(name.to_string());
        if out.len() >= MAX_ARTISTS {
            break;
        }
    }
    out
}

/// Validate a model's answer for one batch: clamp indices into range, coerce
/// anything unparseable to -1, reject a wrong-length answer outright.
fn normalise_singers(line_count: usize, singers: &Value, artist_count: usize) -> Option<Vec<i64>> {
    let arr = singers.as_array()?;
    if arr.len() != line_count {
        return None;
    }
    Some(
        arr.iter()
            .map(|v| match v.as_i64() {
                Some(n) if n >= 0 && (n as usize) < artist_count => n,
                _ => -1,
            })
            .collect(),
    )
}

/// Absorb runs shorter than `min_run` into the run before them, so the mic does
/// not flicker between dancers for a single line.
fn smooth_runs(singers: Vec<i64>, min_run: usize) -> Vec<i64> {
    if singers.is_empty() {
        return singers;
    }
    let mut out = singers;
    let mut start = 0;
    while start < out.len() {
        let mut end = start;
        while end + 1 < out.len() && out[end + 1] == out[start] {
            end += 1;
        }
        let length = end - start + 1;
        if length < min_run && start > 0 {
            let fill = out[start - 1];
            for slot in out.iter_mut().take(end + 1).skip(start) {
                *slot = fill;
            }
        }
        start = end + 1;
    }
    out
}

/// Work out who sings each line. None when there is no usable answer (no
/// provider, a solo artist, failures, or an all-unknown result).
pub fn attribute_lines(cue_texts: &[String], title: &str, artist: &str) -> Option<(Vec<String>, Vec<i64>)> {
    if cue_texts.is_empty() || !llm::is_available() {
        return None;
    }
    let lineup = split_artists(artist);
    if lineup.len() < 2 {
        return None; // one name is not a question
    }

    let mut singers: Vec<i64> = Vec::with_capacity(cue_texts.len());
    let mut i = 0;
    while i < cue_texts.len() {
        let end = (i + BATCH_SIZE).min(cue_texts.len());
        let slice = &cue_texts[i..end];
        let user = format!(
            "Song: {}\nArtists (use these indices):\n{}\n\nLines:\n{}\n\nReturn exactly {} indices.",
            if title.is_empty() { "unknown" } else { title },
            lineup.iter().enumerate().map(|(idx, n)| format!("  {idx}: {n}")).collect::<Vec<_>>().join("\n"),
            slice.iter().enumerate().map(|(idx, t)| format!("{idx}: {}", t.trim())).collect::<Vec<_>>().join("\n"),
            slice.len()
        );
        match llm::convert(SYSTEM_PROMPT, &user, &schema()) {
            Ok(parsed) => match parsed.get("singers").and_then(|s| normalise_singers(slice.len(), s, lineup.len())) {
                Some(mut batch) => singers.append(&mut batch),
                None => singers.extend(std::iter::repeat(-1).take(slice.len())),
            },
            Err(_) => singers.extend(std::iter::repeat(-1).take(slice.len())),
        }
        i = end;
    }

    if singers.iter().all(|s| *s < 0) {
        return None; // every line unknown is the default wearing a hat
    }
    Some((lineup, smooth_runs(singers, 2)))
}

/// Fill in lines the LLM left at -1 (unknown/shared) using the audio's own
/// diarization clusters (ROADMAP.md §5.8, `sidecar/src/diarize.rs`) — the
/// text-based guess above knows the lineup's NAMES but not who is actually
/// singing; diarization knows who is singing but only as anonymous cluster
/// numbers. Combined: for each unresolved line, look at which artist indices
/// the LLM already assigned to OTHER lines sharing this line's diarization
/// cluster, and take the majority; a cluster with no confidently-attributed
/// line anywhere in it is left at -1, same as before.
///
/// `line_cluster` is one diarization cluster id per line — the caller's job
/// to derive from `SpeakerSpan`s and each cue's time range (whichever cluster
/// covers most of the line), since that mapping needs the cue timings this
/// function does not take. `None` for a line diarization said nothing about
/// (e.g. shorter than one embedding window).
///
/// Not yet wired to a live caller: diarization needs raw audio, and today
/// only songs that need transcription (no synced lyrics found) ever have
/// audio in hand at the point attribution runs (`resolve_attribution`
/// fires off lyric text alone, before playback). Wiring this in for an
/// already-synced song needs that capture to happen for attribution's sake
/// too, not only transcription's — a real next step, not a stub.
// Not yet called from anywhere — see the doc comment above for why. Same
// shape as genius.rs's module-level allow: kept ready, not deleted, because
// deleting it would lose the (tested) hard part before the easy wiring step
// exists to use it.
#[allow(dead_code)]
pub fn refine_with_diarization(singers: &[i64], line_cluster: &[Option<u32>]) -> Vec<i64> {
    if singers.len() != line_cluster.len() {
        return singers.to_vec(); // shapes disagree; refuse rather than guess
    }

    // One majority vote per cluster, from lines the LLM already resolved.
    let mut votes: std::collections::HashMap<u32, std::collections::HashMap<i64, u32>> = std::collections::HashMap::new();
    for (&singer, &cluster) in singers.iter().zip(line_cluster) {
        if singer < 0 {
            continue;
        }
        if let Some(c) = cluster {
            *votes.entry(c).or_default().entry(singer).or_default() += 1;
        }
    }
    let majority: std::collections::HashMap<u32, i64> = votes
        .into_iter()
        .filter_map(|(cluster, counts)| counts.into_iter().max_by_key(|(_, n)| *n).map(|(singer, _)| (cluster, singer)))
        .collect();

    singers
        .iter()
        .zip(line_cluster)
        .map(|(&singer, &cluster)| {
            if singer >= 0 {
                return singer;
            }
            cluster.and_then(|c| majority.get(&c).copied()).unwrap_or(-1)
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn split_artists_handles_common_shapes() {
        assert_eq!(split_artists("Drake"), vec!["Drake"]);
        assert_eq!(split_artists("A, B & C"), vec!["A", "B", "C"]);
        assert_eq!(split_artists("Calvin Harris feat. Rihanna"), vec!["Calvin Harris", "Rihanna"]);
        assert_eq!(split_artists("Seedhe Maut x MC"), vec!["Seedhe Maut", "MC"]);
        // Case-insensitive dedupe.
        assert_eq!(split_artists("Drake & drake"), vec!["Drake"]);
    }

    #[test]
    fn normalise_rejects_wrong_length_and_clamps() {
        let v = json!([0, 1, 5, -1, "x"]);
        // artist_count 2 → 5 and "x" become -1.
        assert_eq!(normalise_singers(5, &v, 2), Some(vec![0, 1, -1, -1, -1]));
        assert_eq!(normalise_singers(4, &v, 2), None); // length mismatch
    }

    #[test]
    fn refine_fills_an_unknown_line_from_its_clusters_majority() {
        // Lines 0,1 confidently attributed to artist 0, both in cluster 5.
        // Line 2 is unknown but shares cluster 5 — should inherit artist 0.
        let singers = vec![0, 0, -1];
        let clusters = vec![Some(5), Some(5), Some(5)];
        assert_eq!(refine_with_diarization(&singers, &clusters), vec![0, 0, 0]);
    }

    #[test]
    fn refine_never_overwrites_an_already_confident_line() {
        let singers = vec![0, 1];
        let clusters = vec![Some(9), Some(9)]; // same cluster, disagreeing labels
        assert_eq!(refine_with_diarization(&singers, &clusters), vec![0, 1], "existing LLM answers must not be overwritten");
    }

    #[test]
    fn refine_leaves_a_cluster_with_no_confident_line_at_unknown() {
        let singers = vec![-1, -1];
        let clusters = vec![Some(1), Some(1)];
        assert_eq!(refine_with_diarization(&singers, &clusters), vec![-1, -1]);
    }

    #[test]
    fn refine_leaves_a_line_with_no_cluster_at_unknown() {
        let singers = vec![0, -1];
        let clusters = vec![Some(1), None];
        assert_eq!(refine_with_diarization(&singers, &clusters), vec![0, -1]);
    }

    #[test]
    fn refine_takes_the_majority_when_a_clusters_known_lines_disagree() {
        // Cluster 2 has two lines attributed to artist 0 and one to artist 1;
        // the unknown line should inherit the majority, 0.
        let singers = vec![0, 0, 1, -1];
        let clusters = vec![Some(2), Some(2), Some(2), Some(2)];
        assert_eq!(refine_with_diarization(&singers, &clusters)[3], 0);
    }

    #[test]
    fn refine_refuses_mismatched_lengths_rather_than_guessing() {
        let singers = vec![0, -1];
        let clusters = vec![Some(1)];
        assert_eq!(refine_with_diarization(&singers, &clusters), singers);
    }

    #[test]
    fn smooth_absorbs_short_runs() {
        // A lone 1 between runs of 0 is absorbed into the preceding 0.
        assert_eq!(smooth_runs(vec![0, 0, 1, 0, 0], 2), vec![0, 0, 0, 0, 0]);
        // The first run stands even if short.
        assert_eq!(smooth_runs(vec![1, 0, 0], 2), vec![1, 0, 0]);
    }
}
